Source/WebCore/Modules/mediasession/MediaMetadata.cpp

```diff
@@ -159,22 +161,101 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
     return { };
 }
 
+// We attempt to give a score to the image dimensions. This score is between 0 and 1.
+// A negative score indicates an invalid image
+// A score of 0 is for images smaller than the minimum size
+// A score of 1 indicates an image of ideal size with an aspect ratio of 1 (square)
+// The closer to the ideal size, the higher the score.
+float MediaMetadata::imageDimensionsScore(int width, int height) const
+{
+    IntSize size { width, height };
+    if (size.isEmpty())
+        return -1;
+
+    if (size.maxDimension() <= m_minimumSize)
+        return 0; // Ignore images that are too small.
+
+    // We account for the aspect ratio in scoring the "best" artwork.
+    // We prefer artwork's images with a square AR.
+    double longEdge = size.maxDimension();
+    double shortEdge = size.minDimension();
+    auto aspectRatioCoefficient = shortEdge / longEdge;
+
+    if (longEdge < m_idealSize)
+        return aspectRatioCoefficient * (0.8 * (longEdge - m_minimumSize) / (m_idealSize - m_minimumSize) + 0.2);
+    return aspectRatioCoefficient * (1.0 * m_idealSize / longEdge);
+}
+
 void MediaMetadata::refreshArtworkImage()
 {
-    const auto& mediaImages = m_metadata.artwork;
-    if (mediaImages.isEmpty()) {
-        m_artworkLoader = nullptr;
+    m_artworkImageSrc = String();
+    m_artworkImage = nullptr;
+    m_artworkLoader = nullptr;
+    m_sortedSrc.clear();
+    if (m_metadata.artwork.isEmpty())
         return;
-    }
-    if (!m_session || !m_session->document() || mediaImages[0].src == m_artworkImageSrc)
+    if (!m_session || !m_session->document())
         return;
-    // FIXME: Implement a heuristic to retrieve the "best" image.
-    m_artworkImageSrc = mediaImages[0].src;
-    m_artworkLoader = makeUnique<ArtworkImageLoader>(*m_session->document(), m_artworkImageSrc, [this](Image* image) {
-        if (!image || !image->data())
+
+    // First look into the artwork's sizes attributes to attempt to determine the best score.
+    Vector<IntSize> sizes(m_metadata.artwork.size(), [&](size_t index) -> IntSize {
```

Comment:
Would it be possible to directly create m_sortedSrc without creating this sizes vector?